beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+    if (!grid_cell.cluster_id.has_value()) {
+      grid_cell.cluster_id = next_cluster_id++;
+    }
+
+    // process the neighbors of the cell; if they don't have a cluster id already assigned
+    // then assign them one and add them to the queue with an inflated priority
+    // to ensure they get processed ASAP before moving on to other local peaks.
+    // Notice that with this algorithm each cell will go through the priority queue at most twice.
+
+    const auto get_neighbor_hash = [&](const auto& neighbor_pose_in_representative) {
+      return spatial_hash_function_(representative_pose_in_world * neighbor_pose_in_representative);
+    };
+
+    const auto filter_invalid_neighbors = [&](const auto& neighbor_hash) {
+      auto it = grid_cell_data.find(neighbor_hash);
+      return (
+          (it != grid_cell_data.end()) &&            // is within the map
+          (!it->second.cluster_id.has_value()) &&    // AND not yet mapped to a cluster
+          (it->second.weight <= grid_cell_weight));  // AND has lower weight than the current cell
+    };
+
+    auto valid_neighbor_hashes_view =                     //
+        neighbors |                                       //
+        ranges::views::transform(get_neighbor_hash) |     //
+        ranges::views::filter(filter_invalid_neighbors);  //
+
+    for (const auto& neighbor_hash : valid_neighbor_hashes_view) {
+      auto& neighbor = grid_cell_data[neighbor_hash];
+      neighbor.cluster_id = grid_cell.cluster_id;
+      const auto inflated_priority =
+          weight_cap + neighbor.weight;  // since weights are capped at weight_cap, this gives us a value that is
+                                         // guaranteed to be higher than any other weight from a local maximum.
+      grid_cell_queue.push({inflated_priority, neighbor_hash});  // reintroduce with inflated priority
+    }
+  }
+}
+
+/// @brief For each cluster, estimate the mean and covariance of the states that belong to it.
+/// @tparam StateType Type used for the pose estimation.
+/// @tparam CovarianceType Type used for the covariance estimation.
+/// @tparam GridCellDataMapType Type of the grid cell data map.
+/// @tparam Range Range type of the states.
+/// @tparam Weights Range type of the weights.
+/// @tparam Hashes Range type of the hashes.
+/// @param grid_cell_data Grid cell data map that has already been processed by the clustering algorithm.
+/// @param states Range containing the states of the particles.
+/// @param weights Range containing the weights of the particles.
+/// @param hashes Range containing the hashes of the particles.
+/// @return A vector of tuples, containing the weight, mean and covariance of each cluster, in no particular order.
+template <class StateType, class CovarianceType, class GridCellDataMapType, class Range, class Weights, class Hashes>
+[[nodiscard]] auto
+estimate_clusters(GridCellDataMapType&& grid_cell_data, Range&& states, Weights&& weights, Hashes&& hashes) {
```

Comment:
Can we change the signature to something like this?

```cpp
auto estimate_clusters(Range&& states, Weights&& weights, Clusters&& clusters)
```

I think this is a good candidate for `range_statistics.hpp` if we remove the `GridCellDataMapType` dependency.

I played a bit with an alternative implementation that doesn't remove the dependency but shows how it's a trivial step forward: https://github.com/Ekumen-OS/beluga/commit/a87ce57caf62c7ca9d713847d1a464f7e3c8be79 WDYT?

Reply:
Good idea. Done.